File: src/bin/mt/apksign/data/ByteArrayDataSource.java
package bin.mt.apksign.data;

import java.io.EOFException;
import java.io.IOException;
import java.io.OutputStream;

public class ByteArrayDataSource implements DataSource {
    private byte[] data;
    private int start;
    private int size;
    private int pos;

    ByteArrayDataSource(byte[] data, int start, int size) {
        if (start + size > data.length)
            throw new IllegalArgumentException();
        this.data = data;
        this.start = start;
        this.size = size;
        this.pos = 0;
    }

    @Override
    public long size() {
        return size;
    }

    @Override
    public long pos() {
        return pos;
    }

    @Override
    public void reset() {
        pos = 0;
    }

    @Override
    public void copyTo(OutputStream os, long length) throws IOException {
        if (length > remaining())
            throw new EOFException();
        os.write(data, start + pos, (int) length);
        pos += length;
    }

    public byte[] getBuffer() {
        return data;
    }

    public int getStart() {
        return start;
    }
}

File: Android/app/src/main/java/bin/mt/apksign/data/ByteArrayDataSource.kt
package bin.mt.apksign.data

import java.io.EOFException
import java.io.IOException
import java.io.OutputStream

class ByteArrayDataSource internal constructor(data: ByteArray, start: Int, size: Int) :
    DataSource {
    val buffer: ByteArray
    val start: Int
    private val size: Int
    private var pos: Int

    init {
        require(start + size <= data.size)
        buffer = data
        this.start = start
        this.size = size
        pos = 0
    }

    override fun size(): Long {
        return size.toLong()
    }

    override fun pos(): Long {
        return pos.toLong()
    }

    override fun reset() {
        pos = 0
    }

    @Throws(IOException::class)
    override fun copyTo(os: OutputStream, length: Long) {
        if (length > remaining()) throw EOFException()
        os.write(buffer, start + pos, length.toInt())
        pos += length.toInt()
    }
}